### ISSUE TITLE
Move mksurfdata_esmf CTSM5.2 python support scripts gen_mksurfdata_*.py to python directory to get python development support

### ISSUE THREAD

**Issue by ekluzek** [ESCOMP/CTSM#2191]
The gen_mksurfdata_*.py support scripts for mksurfdata_esmf in ctsm5.2 are all separate scripts under the tools/mksurfdata_esmf directory. As with other python scripts in order to use other python development tools, I'll be moving the core to the python subdirectory. Then for development we automatically get python linting as well as black formatting, and it facilitates adding unit testing for the scripts. This will improve the current development as well as future development. Since, these scripts are core to the support of surface dataset creation it's critical to have support for them.

Also as part of that the gen_mksurfdata_multi and gen_mksurfdata_single scripts have code duplication that makes them hard to support and requires separate testing for them. As part of this I'll reduce that duplication that's hard to support by having the multi script be based on top of the single script with additional functionality to support different standard resolution/scenarios to be built.

The other thing to fix is that there is a subprocess.run from the jobscript scripts to the namelist one. I've been having similar problems with it as we have in other contexts. By moving the guts of the python scripts to the python directory, I'll be able to avoid using a subprocess.run and fix that issue as well. Strangely the subprocess.run can sometimes work fine, and then unrelated changes seem to make it break. But in any case this will make the use of the scripts more robust.

I thought about continuing to fight with the scripts as they are, but this will improve current and future development. And by focussing on the minimum changes to make do this it will improve the current development.

@slevis-lmwg and I discussed this earlier today.

**Comment by ekluzek** [comment]
Future issues for this would be to add more extensive unit tests and modularizing these scripts more. As a first step I'm just planning on doing what's needed to make this work reasonable to do. This also allows this issue to be something that can be completed right away.

**Comment by mvertens** [comment]
@ekluzek @slevis-lmwg - what is the timeline for getting this new surface dataset generation into the main CTSM development. Is it necessary to wait for all of these proposed changes before bringing it in and then incrementally adding the proposed new features. There is significant interest in the community to using this new capability - so I am just wondering what the priority is for getting into the main development code as soon as possible.

**Comment by wwieder** [comment]
Fair question, @mvertens.  Currently the SE team is working on finalizing the updates to roughness length calculations and snow albedo.  They're also working on a new dust emission scheme, crop model developments, SLIM, developing a workflow to run tower sites for PLUMBER2, modifications to how we handle time on history files, and running experiments to evaluate alternative snow thermal conductivity and other parameter changes to try and increase plant survival in Arctic systems (and likely more I'm overlooking).  These have all been deemed high priority for CESM3 developments.  In short we're doing our best, and trying to prioritize too many activities.

For the 5.2 tag our plan was for this to be a very short development series for CTSM, with only one or two tags that introduce the capability and datasets before opening up the CTSM5.3dev line that builds on this infrastructure. 

I'm traveling this week and haven't been able to check in with Erik and Sam, but it seems like this issue is intended to increase the testing and maintainability of the scripts we're developing.  All of this said, it's still a fair question, @ekluzek and @slevis-lmwg, when do you expect we'll be able to make the 5.2 tag?

**Comment by mvertens** [comment]
@wwieder - thanks for the clarifications. That's very helpful to know.

**Comment by wwieder** [comment]
@mvertens I also wondered if one of the preliminary surface datasets we've already created internally for CESM development would be helpful for the NorESM group in the interim?   I think @ekluzek likely has already created a number of standard grids that may be of use?

**Comment by mvertens** [comment]
@wwieder - actually the interest is in having the new parallel utility for generating surface datasets. Having an idea of when that will be available in the main development code would be really helpful.

**Comment by ekluzek** [comment]
@mvertens and @wwieder here is an update on timelines.

The project board for CTSM5.2 is here:

https://github.com/ESCOMP/CTSM/projects/36

It looks more complicated than it really should because we ended up with multiple cards that really describe the same thing. That said -- three important things. First is that we just finalized (yesterday) the list of grids that we are going to support. Second is that a small group is meeting to talk about the project next Tuesday. Note, @mvdebolskiy is joining us for that meeting he'll be able to update NorESM folks afterwards, as well as represent their interests. Third the list of important features that need to come in are:

- Finish the Makefile to easily make all datasets (I'm working on this till it's done)
- Handle ocean separate from wetland (@slevis-lmwg is working on)
- Create a bash script that creates a landuse.timeseries file for the smallvilleIA test cases
- Port the process to Derecho (this is important as we need to create all datasets on Derecho, so we have the possibility of replicating answers moving forward)

The list of grids we will support out of the box (that we just finalized) is in this document...

https://docs.google.com/spreadsheets/d/1Osq56e423CF107zhoNQ0VS7-iH_JXLF9AtCvBdXyfJ4

@wwieder @slevis-lmwg created most of the previous standard resolutions, and I've created a version of some of the oddball resolutions. We'll also have a complete set of preliminary files soon. There's probably going to a few versions of preliminary versions before we get to the final ones.

Happy to answer any other questions regarding above. Hopefully, that helps give the big picture of what's going on...
